In [4]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.cluster import DBSCAN
from sklearn.preprocessing import StandardScaler

In [16]:
df = pd.read_csv('D:/Data Science/Assignment/07Clustring/crime_data.csv')
df.head()


Unnamed: 0  Murder  Assault  UrbanPop  Rape
0     Alabama    13.2      236        58  21.2
1      Alaska    10.0      263        48  44.5
2     Arizona     8.1      294        80  31.0
3    Arkansas     8.8      190        50  19.5
4  California     9.0      276        91  40.6

In [21]:
#find the missing values
print(df.shape)
df.isna().sum()

(50, 5)


Unnamed: 0    0
Murder        0
Assault       0
UrbanPop      0
Rape          0
dtype: int64

there is no missing value so no need of imputation

In [30]:
#scaling 
x= df.iloc[:,1:].values
sc_x=StandardScaler()
x=sc_x.fit_transform(x)

In [33]:
#create the model
model=DBSCAN(eps=1.1,min_samples=3)
model.fit(x)

DBSCAN(eps=1.1, min_samples=3)

In [35]:
model.labels_

array([ 0, -1,  1,  2, -1, -1,  2, -1, -1,  0,  2,  2,  1,  2,  2,  2,  2,
        0,  2,  1,  2,  1,  2,  0,  2,  2,  2, -1,  2,  2,  1,  1,  0,  2,
        2,  2,  2,  2,  2,  0,  2,  0,  1,  2,  2,  2,  2,  2,  2,  2],
      dtype=int64)

In [38]:
df['cluster']=model.labels_

In [40]:
df.cluster.value_counts()

 2    30
 1     7
 0     7
-1     6
Name: cluster, dtype: int64

In [42]:
df1=df.groupby('cluster').agg('mean')

In [44]:
df1.sort_values('cluster')

Murder     Assault   UrbanPop       Rape
cluster                                             
-1       10.066667  261.333333  75.000000  36.250000
 0       14.671429  251.285714  54.285714  21.685714
 1       11.014286  262.571429  77.142857  28.800000
 2        4.973333  112.433333  63.566667  16.356667

In [51]:
def range_x(x):
    return(np.min(x),np.max(x))
for i in range(1,(df1.shape[1])+1):
    print(df.columns[i],range_x(df.iloc[:,i:i+1][df.cluster==2]))
    print()

Murder (Murder    0.8
dtype: float64, Murder    9.7
dtype: float64)

Assault (Assault    45
dtype: int64, Assault    190
dtype: int64)

UrbanPop (UrbanPop    32
dtype: int64, UrbanPop    89
dtype: int64)

Rape (Rape    7.3
dtype: float64, Rape    29.3
dtype: float64)



# Inferences

we reduce eps value the number of outlier will increases and we increase the minimum samples beyond 5 no of outliers will increases. so optimum value of eps is 1.1 and minimum samples is 3. we get result from this data we could not get best combination, because in one cluster we have 30 data points which is 60% of total dataset. but noice data points are in limit that is about 14% of total dataset

we should not use DBSCAN for this data set.

# Insights

for this data we have maximum 2 cluster and in murder have  min = 0.8 and max is 9.7, in assault have min= 45 and max = 190 ,in urbanpop have min=32 and man=89 , in rape have min=7.3 and max=29.3